### PR TITLE
Creating methods to render geoJson on map

### PR DESCRIPTION
**Pull request by corrigancd**
part of https://github.com/sirensolutions/kibi-internal/issues/10600
close https://github.com/sirensolutions/kibi-internal/issues/10616

Note there are new files in the directives folder. This is a UI change which is a start to adding WFS overlays and is not necessary for this actual issue, for this reason this should not be merged. 

**Exposed methods for calling from script**
renderScriptingGeoJson(layerName, geoJson) - is the one to call from script with the specified arguments **(currently called from the ES response watcher)**
getGeoBoundingBox() - will return the current map extent for use in script

There are four testing geoJson files in **public/testGeoJson/**. The one ending in EU is used in example 1 below. The one ending in World contains many multipolygon types, example 2 below. The other two are miscellaneous. 

**Example 1**: 
FilteringByPolygon

**Example 2**:
image

I would recommend looking at //TODO and console.log comments in callbacks.js for a description of the current situation, in short, polygons without holes are supported, multipolygons and polygons with holes will require adjustment of the geofilter.js file.

### CODE REVIEW

**Comment by jefersonzanim** [comment]
Looks good. :)

**Comment by corrigancd** [comment]
Multipolygons and polygons with donuts now handled. There are still WFS related material here and shouldn't be merged yet: 
filtering and combining multiple geofilters

The following has been tested manually:
WRITING FILTERS:
drawn polygons Y
rectangles Y
simple vector polygons Y
simple vector polygons with donuts Y (applied filters work but don't draw the donut)
multipolygon vector Y
multipolygon vector with donuts Y (applied filters work but don't draw the donut)
circles Y (applied filters work but don't draw for circles #10803 created)


COMBINING FILTERS: 
INITIALLY DRAW:
drawn polygons
rectangles
simple vector polygons
simple vector polygons with donuts
multipolygon vector
multipolygon vector with donuts
circles


THEN COMBINE THE INITIALLY DRAWN WITH:
drawn polygons Y
rectangles Y
simple vector polygons Y
simple vector polygons with donuts Y
multipolygon vector Y
multipolygon vector with donuts Y
circles Y

**Comment by jccq** [comment]
Edwin  looks great can you do a gif where you show activating the extra
layers first and the using them as filters .

eg normal map then activate the layer with polygon for example the
rectangle then activate the layer with the council boundaries then click on
the two

could you please try this would work correctly 4 layers were for example
you have all the boundaries of all the countries

So click on Italy Germany etc


 can you please check that it works

Thanks

**Comment by jccq** [comment]
Ok can you pls send me a gif which shows you activating the JSON layers and
then clicking on them to create filters sort of like the first gif but
starting from a normal map then activating layers which are clickable
Thanks

**Comment by corrigancd** [comment]
I call the exposed method via the ES response watcher just for testing. I don't think you want a GIF of me hard coding that into the vis controller and there is currently no other way to do it.

**Comment by jccq** [comment]
Ah ok sorry I thought this was showing a layer of polygons so you could
just activate the layer  not that you were injecting a polygon with no
layer

Thanks

File: public/directives/wfsOverlay.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            //if there is a valid response from WFS server
+            if (wfsLayers) {
+              scope.layer.wfsLayers = doWmsToUiSelectFormat(wfsLayers);
+              if (scope.layer.layers) {
+                scope.layer.selected = findSelectedLayer(scope.layer.wfsLayers, scope.layer.layers);
+              } else {
+                scope.layer.selected = [];
+              }
+
+              scope.layer.wfsCapabilitiesSwitch = 1;
```
Comment (szydan) [inline public/directives/wfsOverlay.js]: if the possible values are only 0 or 1 I prefer to use boolean true, false here

File: public/directives/wfsOverlays.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      scope: {
+        layers: '='
+      },
+      template: require('./wfsOverlays.html'),
+      link: function (scope, element, attrs) {
+
+        scope.addLayer = function () {
+          if (!scope.layers) scope.layers = [];
+          scope.layers.push({
+            // minZoom: 13,
```
Comment (szydan) [inline public/directives/wfsOverlays.js]: remove commented code

Comment (szydan) [reply public/directives/wfsOverlays.js]: It looks a bit weird that a layer is composed from minZoom and color 
Maybe call it "layer properties"

File: public/vislib/_map.js [hunk 3]
```diff
@@ -222,6 +224,30 @@ define(function (require) {
...
+        layer.destroy();
+        self._layerControl.removeLayer(layer);
+        self.map.removeLayer(layer);
+      });
+      this._vectorOverlays = {};
+      if (this._toolbench) this._toolbench.removeTools();
+    };
+
+    TileMapMap.prototype.clearWfsOverlays = function () {
+      const self = this;
```
Comment (szydan) [inline public/vislib/_map.js]: You use arrow function here so the self should not be needed

File: public/vislib/_map.js [hunk 4]
```diff
@@ -222,6 +224,30 @@ define(function (require) {
       if (this._toolbench) this._toolbench.removeTools();
     };
 
+    TileMapMap.prototype.clearVectorLayers = function () {
+      const self = this;
+      Object.keys(this._vectorOverlays).forEach(function (key) {
```
Comment (szydan) [inline public/vislib/_map.js]: If you use arrow function you could remove the "self"

File: public/vislib/geoFilter.js [hunk 5]
```diff
@@ -13,44 +13,140 @@ define(function (require) {
...
     }
 
-    function _applyFilter(newFilter, field, indexPatternName, _sirenMeta) {
-      let numFilters = 1;
-      if (_.isArray(newFilter)) {
-        numFilters = newFilter.length;
-        newFilter = {
-          bool: {
-            should: newFilter
+    function _analyseMultiPolygon(polygons, field) {
```
Comment (szydan) [inline public/vislib/geoFilter.js]: would be good to have a unit test for
 _analyseMultiPolygon
and 
_analyseSimplePolygon

to be sure that they always produce correct objects

File: public/vislib/geoFilter.js [hunk 6]
```diff
@@ -13,44 +13,140 @@ define(function (require) {
...
         negate: false,
         index: indexPatternName,
-        _siren: _sirenMeta,
         key: field
       };
       queryFilter.addFilters(newFilter);
-    }
+    };
 
     function _combineFilters(newFilter, existingFilter, field) {
```
Comment (szydan) [inline public/vislib/geoFilter.js]: Would be nice to have a unit test to be sure we combine in correct way

File: public/vislib/geoFilter.js (hunk 5 above)
Comment (corrigancd) [reply public/vislib/geoFilter.js]: Issue created

File: public/vislib/_map.js (hunk 4 above)
Comment (corrigancd) [reply public/vislib/_map.js]: I've tried the below and it doesn't work:

```
    TileMapMap.prototype.clearVectorLayers = () => {
      Object.keys(this._vectorOverlays).forEach(function (key) {
        const layer = this._vectorOverlays[key];
        layer.destroy();
        this._layerControl.removeLayer(layer);
        this.map.removeLayer(layer);
      });
      this._vectorOverlays = {};
      if (this._toolbench) this._toolbench.removeTools();
    };
```

File: public/directives/wfsOverlays.js (hunk 2 above)
Comment (corrigancd) [reply public/directives/wfsOverlays.js]: I think it's a default value for what will ultimately become a layer

File: public/directives/wfsOverlay.js (hunk 1 above)
Comment (corrigancd) [reply public/directives/wfsOverlay.js]: Done

File: public/vislib/_map.js (hunk 4 above)
Comment (szydan) [reply public/vislib/_map.js]: you need an arrow function here 
forEach(function (key) {

Comment (corrigancd) [reply public/vislib/_map.js]: ah ok,

**Pull request merged**